### PR TITLE
tests: use gojq - part 1

### PR DESCRIPTION
**Pull request by bboozzoo**
https://github.com/itchyny/gojq is almost-compatible replacement for jq, with additional features such as native support for YAML input and output. 

Attempt to switch the tests to use gojq instead of jq. Where appropriate drop use `remarshal`, since YAML support is built in.

Related: SNAPDENG-33306

### CODE REVIEW

File: tests/main/aux-info/task.yaml [hunk 1]
```diff
@@ -9,17 +9,16 @@ systems: [ubuntu-18.04-64, ubuntu-2*, ubuntu-core-*, fedora-*]
 
 prepare: |
   snap install snap-store
-  snap install jq
 
 execute: |
-  snap_id=$(snap info snap-store | grep snap-id | awk '{ print $2 }')
-  jq --sort-keys .media < "/var/cache/snapd/aux/${snap_id}.json" > media.json
+  snap_id=$(snap info snap-store | gojq -r --yaml-input '.["snap-id"]')
+  gojq .media < "/var/cache/snapd/aux/${snap_id}.json" > media.json
```
Comment (valentindavid) [inline tests/main/aux-info/task.yaml:15]: Does gojq always sort keys?

Comment (valentindavid) [reply tests/main/aux-info/task.yaml:15]: I just saw your comment somewhere else. Nevermind.

File: tests/lib/prepare-restore.sh [hunk 2]
```diff
@@ -608,6 +608,15 @@ prepare_project() {
...
     fi
+
+    # native jq replacement, but still with some incompatibilies, see
+    # https://github.com/itchyny/gojq
+    # major differences:
+    # - map keys are sorted by default
+    # - with --yaml-input, can parse YAML
+    GOBIN=$PROJECT_PATH/tests/bin \
+    CGO_ENABLED=0 \
+        go install github.com/itchyny/gojq/cmd/gojq@v0.12.16
```
Comment (zyga) [inline tests/lib/prepare-restore.sh:619]: We might consider just pulling one of the pre-built artefacts instead:

- https://github.com/itchyny/gojq/releases/download/v0.12.16/gojq_v0.12.16_linux_amd64.tar.gz
- https://github.com/itchyny/gojq/releases/download/v0.12.16/gojq_v0.12.16_linux_arm64.tar.gz

The build could be only done on arch mismatch.

File: tests/main/interfaces-custom-device-app-slot/task.yaml [hunk 3]
```diff
@@ -28,6 +28,8 @@ prepare: |
     snap debug can-manage-refreshes | MATCH false
 
     snap ack "$TESTSLIB/assertions/testrootorg-store.account-key"
+    snap ack "$TESTSLIB/assertions/developer1.account"
+    snap ack "$TESTSLIB/assertions/developer1.account-key"
```
Comment (zyga) [inline tests/main/interfaces-custom-device-app-slot/task.yaml:32]: How is this related?

File: tests/main/snap-disconnect/task.yaml [hunk 4]
```diff
@@ -13,13 +13,12 @@ prepare: |
...
     snap pack "$TESTSLIB"/snaps/home-consumer
     snap install --dangerous "$SNAP_FILE"
-    snap install --edge jq
 
 execute: |
     inspect_connection() {
       CONN="$1"
-      # shellcheck disable=SC2002
-      cat /var/lib/snapd/state.json | jq --arg CONN "$CONN" -r '.data["conns"] | has($CONN)'
+      # shellcheck disable=SC2002,SC2016
```
Comment (zyga) [inline tests/main/snap-disconnect/task.yaml:20]: What is 2016?

File: tests/main/store-state/snap/meta/snap.yaml.in [hunk 5]
```diff
@@ -1,6 +1,7 @@
 name: SNAPNAME
 summary: generic snap
 version: '1.0'
+base: core22
```
Comment (zyga) [inline tests/main/store-state/snap/meta/snap.yaml.in:4]: Could we add this in the other pass later?

File: tests/main/store-state/task.yaml [hunk 6]
```diff
@@ -40,6 +29,10 @@ execute: |
         snap info core | MATCH "store-url:.*https://snapcraft.io"
     fi
 
+    # install dependecies before switching to fake store
```
Comment (zyga) [inline tests/main/store-state/task.yaml:32]: This seems related to the base change, but I don't really understand the significance. Can we add this in a second pass later?

**Review by zyga — COMMENTED** [review]
First pass mostly looks good. I left a few questions about things that seemed like they could be broken out to another polar quest later.

File: tests/main/store-state/task.yaml (hunk 6 above)
Comment (bboozzoo) [reply tests/main/store-state/task.yaml:32]: Not really, fake store does not forward install requests to the real store. The test snap installed here had no base, hence defaulted to 'core'. We installed jq & remarshal both of which pulled in 'core'. Since the install happened before switching over to fake store, it all worked by accident as the dependencies happened to be satisfied. Without jq or remarshal nothing pulls in the dependencies. That's why the snap's base was switched to core22 too, just to avoid pulling in core.

File: tests/main/snap-disconnect/task.yaml (hunk 4 above)
Comment (bboozzoo) [reply tests/main/snap-disconnect/task.yaml:20]: https://www.shellcheck.net/wiki/SC2016 I have no clue why it wasn't picked up before due to `-r '... | has ($CONN)'`

File: tests/lib/prepare-restore.sh (hunk 2 above)
Comment (bboozzoo) [reply tests/lib/prepare-restore.sh:619]: are those built with CGO disabled?

File: tests/main/interfaces-custom-device-app-slot/task.yaml (hunk 3 above)
Comment (bboozzoo) [reply tests/main/interfaces-custom-device-app-slot/task.yaml:32]: it never worked, but store-state was running without -e, so it failed quietly on https://github.com/canonical/snapd/pull/14686/files/08eab5ab589736cc661f00cffec08f72bcf1f0a7#diff-234265efe637e5730e201767212cce6ce6f4f5fea39bfb026a01913fa281afccL128

File: tests/main/store-state/snap/meta/snap.yaml.in (hunk 5 above)
Comment (zyga) [reply tests/main/store-state/snap/meta/snap.yaml.in:4]: @bboozzoo ?

**Review by zyga — COMMENTED** [review]
Two comments:

Let's try to separate changes to bases to another PR so that it is clear they are not a factor here. Use core if you have to have something.

The store endpoint for find is sometimes throttled, perhaps we can use info instead?

File: tests/main/store-state/snap/meta/snap.yaml.in (hunk 5 above)
Comment (bboozzoo) [reply tests/main/store-state/snap/meta/snap.yaml.in:4]: I'd prefer to avoid pulling in core because it masked bugs and wrong assumptions in the tests.

File: tests/nested/manual/core20-install-device-file-install-ubuntu-save-via-hook/task.yaml [hunk 7]
```diff
@@ -78,12 +73,10 @@ prepare: |
     cp prepare-device pc-gadget/meta/hooks/prepare-device
 
     echo "Add the extra hooks definition to the snap.yaml"
-    # convert our yaml files to json (yaml is easier to write and maintain in VCS)
-    yaml2json < pc-gadget/meta/snap.yaml | tee pc-gadget/meta/snap.json > /dev/null
-    yaml2json < snap-yaml-extras.yaml | tee snap-yaml-extras.json > /dev/null
-    # slurp our two json files together into one json document, then convert 
+    # slurp our two json files together into one json document, then convert
     # back to yaml and write out to the snap.yaml in the unpacked gadget snap
```
Comment (olivercalder) [inline tests/nested/manual/core20-install-device-file-install-ubuntu-save-via-hook/task.yaml:77]:
```suggestion
    # slurp our two yaml files together into one yaml document, then write
    # out to the snap.yaml in the unpacked gadget snap
```

File: tests/nested/manual/core20-install-device-file-install-ubuntu-save-via-hook/task.yaml [hunk 8]
```diff
@@ -78,12 +73,10 @@ prepare: |
...
-    jq -s '.[0] * .[1]' <(cat snap-yaml-extras.json) <(cat pc-gadget/meta/snap.json) | json2yaml | tee pc-gadget/meta/snap.yaml
+    gojq -s --yaml-input --yaml-output '.[0] * .[1]' <(cat snap-yaml-extras.json) <(cat pc-gadget/meta/snap.json) | \
```
Comment (olivercalder) [inline tests/nested/manual/core20-install-device-file-install-ubuntu-save-via-hook/task.yaml:78]:
```suggestion
    gojq -s --yaml-input --yaml-output '.[0] * .[1]' <(cat snap-yaml-extras.yaml) <(cat pc-gadget/meta/snap.yaml) | \
```

File: tests/nested/manual/core20-install-device-file-install-via-hook-hack/task.yaml [hunk 9]
```diff
@@ -79,12 +74,10 @@ prepare: |
     cp prepare-device pc-gadget/meta/hooks/prepare-device
 
     echo "Add the extra hooks definition to the snap.yaml"
-    # convert our yaml files to json (yaml is easier to write and maintain in VCS)
-    yaml2json < pc-gadget/meta/snap.yaml | tee pc-gadget/meta/snap.json > /dev/null
-    yaml2json < snap-yaml-extras.yaml | tee snap-yaml-extras.json > /dev/null
-    # slurp our two json files together into one json document, then convert 
+    # slurp our two json files together into one json document, then convert
     # back to yaml and write out to the snap.yaml in the unpacked gadget snap
```
Comment (olivercalder) [inline tests/nested/manual/core20-install-device-file-install-via-hook-hack/task.yaml:78]:
```suggestion
    # slurp our two yaml files together into one taml document, then write
    # out to the snap.yaml in the unpacked gadget snap
```

**Review by olivercalder — CHANGES_REQUESTED** [review]
Thanks! I think I spotted a few typos, but otherwise looks good.

**Review by maykathm — APPROVED** [review]
Nice work

**Comment by bboozzoo** [comment]
I've rebased the branch on top of https://github.com/canonical/snapd/pull/14725

**Review by olivercalder — APPROVED** [review]
Looks good to me, thanks!

File: tests/lib/prepare-restore.sh (hunk 2 above)
Comment (sergiocazzolato) [inline tests/lib/prepare-restore.sh:619]: @bboozzoo what if we create a gojq snap and then we use an alias to keep using jq command?
If we do that, we don't need to update the tests.

**Comment by bboozzoo** [comment]
Failures:
- google:ubuntu-20.04-64:tests/main/interfaces-microstack-support - unrelated
- google-distro-1:debian-11-64:tests/main/registries-cross-config - unrelated
- google*:ubuntu-*-64:tests/main/apparmor-prompting-integration-tests:* - unrelated
- google-nested:ubuntu-20.04-64:tests/nested/manual/snapd-removes-vulnerable-snap-confine-revs:*
bunch of unrelated UC24 nested:
 - google-nested:ubuntu-24.04-64:tests/nested/core/core20-kernel-reseal
- google-nested:ubuntu-24.04-64:tests/nested/manual/core20-early-config
- google-nested:ubuntu-24.04-64:tests/nested/manual/core20-initramfs-time-moves-forward
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-core:install_optional_all
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-core:install_optional_snap
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-core:plain
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-real:partial
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-real:plain
- google-nested:ubuntu-24.04-64:tests/nested/manual/remodel-min-size
- google-nested:ubuntu-24.04-64:tests/nested/manual/core20-4k-sector-size:logical
- google-nested:ubuntu-24.04-64:tests/nested/manual/devmode-snap-seeded-dangerous
- google-nested:ubuntu-24.04-64:tests/nested/manual/hybrid-remodel

**Review by zyga — APPROVED** [review]
Looks good to me.

Thank you for pushing this forward.
